File: 1.13.c
#include <stdio.h>

/*
 *	print histogram of the length of words in input
 *
 *	TODO: print vertically
 */

#define IN  1
#define OUT 0

int main()
{
	int c, i, length, state;
	int lengths[10];
	float n, longest;

	state = OUT;
	length = 0;
	for(i = 0; i < 10; ++i)
		lengths[i] = 0;

	while ((c = getchar()) != EOF)
	{
		if(c == ' ' || c == '\n' || c == '\t')
		{
			if(state == IN) {
				if(length < 10)
					lengths[length]++;
				else
					lengths[9]++;

				length = 0;
			}
			state = OUT;
		}
		else
		{
			++length;
			if (state == OUT)
				state = IN;
		}
	}

	for(i = 1, longest = 0; i < 10; ++i)
		if(lengths[i] > longest)
			longest = lengths[i];


	for(i = 1; i < 10; ++i)
	{
		printf("%d", i);

		if(i == 9)
			printf("+ ");
		else
			printf("  ");

		for(n = 0; n < lengths[i]; n = n + (longest*longest / (25*lengths[i])))
			printf("-");

		printf("\n");
	}
}

File: 1.13.py
#!/usr/bin/env python3

# print histogram of word length of input
# TODO: print vertically

from fileinput import input
from sys import stdout
from numpy import zeros

in_word = False
length  = 0
names = ("1 ", "2 ", "3 ", "4 ", "5 ", "6 ", "7 ", "8 ", "9+")
lengths = zeros(10)
longest = 0

def histogram( names, freqs ) :
    most_freq = 0
    i = 0
    while i < len(freqs):
        if freqs[i] > most_freq:
            most_freq = freqs[i]
        i += 1
    i = 0
    while i < len(names):
        stdout.write(names[i])
        stdout.write(' ')
        n = 0
        while n < freqs[i]:
            stdout.write('-')
            n += pow(most_freq, 2) / (25 * freqs[i])
        stdout.write('\n')
        i += 1
    return


for line in input():
    for c in line:
        if c in (' ', '\t', '\n'):
            if in_word:
                if length < 9:
                    lengths[length] += 1
                else:
                    lengths[9] += 1
            length = 0
            in_word = False
        else:
            length += 1
            in_word = True
lengths = tuple(lengths[1:10])
histogram(names, lengths)
